ProjectSailor/Source/ProjectSailor/InteractionComponent.cpp

```diff
@@ -0,0 +1,184 @@
+#include "Kismet/KismetSystemLibrary.h"
+#include "Kismet/GameplayStatics.h"
+
+
+// Sets default values for this component's properties
+UInteractionComponent::UInteractionComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = true;
+    //reference to world
+	World = GetWorld();
+
+	
+}
+
+
+// Called when the game starts
+void UInteractionComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	StartInterface();
+	// widget->SetVisibility(ESlateVisibility::Hidden);
+	// ...
+	
+}
+
+
+// Called every frame
+void UInteractionComponent::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
+	//calls all the time to perform raycast
+	
+	// ...
+}
+
+
+
+
+void UInteractionComponent::PerformRaycast()
+{
+	//gets a reference of player
+	APlayerController* PlayerController = Cast<APlayerController>(GetOwner()->GetInstigatorController());
+
+	// //gets a reference of Player proyecto
+	auto* player = Cast<AProjectSailorCharacter>(UGameplayStatics::GetPlayerController(GetWorld(), 0)->GetPawn());
+
+	//gets a reference of
+	if (!PlayerController) return;
+
+	//if the player exists we make a reference of cameraLocation and Rotation
+	FVector CameraLocation;
+	FRotator CameraRotation;
+
+	//returns the point of the view of the player on the variables cameraLocation and rotation
+	PlayerController->GetPlayerViewPoint(CameraLocation, CameraRotation);
+
+	// Add a vertical offset to the camera location
+	CameraLocation.Z += 50.f;
+
+	//calculate the end of the raycast and makes it unitary
+	FVector RaycastEnd = CameraLocation + CameraRotation.Vector() * 3000;
+
+	//create hitResult
+	FHitResult HitResult;
+
+	//creates the variable
+	FCollisionQueryParams CollisionParams;
+	CollisionParams.AddIgnoredActor(player);
+
+	// Define the type of collision that is pawn
+	ECollisionChannel TraceChannel = ECC_Pawn;
+
+	// Converti to a valid argument of the LineTraceSingle
+	ETraceTypeQuery TraceTypeQuery1 = UEngineTypes::ConvertToTraceType(TraceChannel);
+
+	//Make parameters of collision
+	bool bHit = UKismetSystemLibrary::LineTraceSingle(World, CameraLocation, RaycastEnd, TraceTypeQuery1, true, { player }, EDrawDebugTrace::ForDuration, HitResult, true);
+
+	//if raycast hit
+	if (bHit)
+	{
+		//the actor that has hit the raycast
+		AActor* HitActor = HitResult.GetActor();
+
+		UDialogueComponentNPC* DialogueComponent = Cast<UDialogueComponentNPC>(HitActor->GetComponentByClass(UDialogueComponentNPC::StaticClass()));
+		if (DialogueComponent)
+		{
+			// widget->SetVisibility(ESlateVisibility::Visible);
+
+			// control if the key E is pressed
+			if (pressedE)
+			{
+				//interact with HitResult
+				InteractNPC(DialogueComponent);
+				pressedE = false;
+			}
+		}
```

Comment:
I think it is more legible if it is separate, because there is not 2 connected conditions as it can be a != null + other condition.
And the reason there is comments on the widgets it is because we will implement them later because it is necessary.

Reply:
Add them in a TODO then, so we can find it in another moment.

`//TODO: Implement this`

Reply:
Done!